recipes/map-over-sequance-of-elements-inside-list.md

```diff
@@ -0,0 +1,73 @@
+# Map over n consecutive elements inside a list
+
+## Problem
+
+I have a list of elements, and I want to map using a function when on
+each loops the function gets n items from list.
+
+e.g.:
+
+for list `(1 2 3 4 5)` and `N = 2` it will call:
+
+```scheme
+(fn 1 2)
+(fn 2 3)
+(fn 3 4)
+(fn 4 5)
+(fn 5)
+```
+
+for list '(1 2 3 4 5) and `N = 3` it will call:
+
+```scheme
+(fn 1 2 3)
+(fn 2 3 4)
+(fn 3 4 5)
+(fn 4 5)
+(fn 5)
+```
+
+and collect the results into single list.
+
+## Solution
+
+The code use function:
+* `take` from recipe: [Select first n elements from list](/select-first-n-elements-from-list/)
+
+```scheme
+(define (seq-map n fn seq-list)
+  (let loop ((seq-list seq-list) (result '()))
+    (if (null? seq-list)
```

Comment:
Is it intentional that `seq-map` will call `fn` with lists shorter than `n`?

```Scheme
(define (take lst n)
  (let loop ((result '()) (i n) (lst lst))
    (if (or (null? lst) (<= i 0))
        (reverse result)
        (loop (cons (car lst) result) (- i 1) (cdr lst)))))

(define (seq-map n fn seq-list)
  (let loop ((seq-list seq-list) (result '()))
    (if (null? seq-list)
        (reverse result)
        (let ((next-list (take seq-list n)))
          (loop (cdr seq-list) (cons (apply fn next-list) result))))))

(define (seq-map/alternative n fn seq-list)
  (define (maybe-take list n)
    (let loop ((n n) (list list) (result '()))
      (cond ((<= n 0) (reverse result))
            ((not (pair? list)) #f)
            (else (loop (- n 1) (cdr list) (cons (car list) result))))))
  (let loop ((seq-list seq-list) (result '()))
    (let ((next-list (maybe-take seq-list n)))
      (if (not next-list)
          (reverse result)
          (loop (cdr seq-list) (cons (apply fn next-list) result))))))

(define (disp . args)
  (for-each display args)
  (newline))

(let ((example '(1 2 3 4 5)))
  (seq-map 3 disp example)
  (newline)
  (seq-map/alternative 3 disp example))
```

displays:

```
123
234
345
45
5

123
234
345
```

Reply:
Yes, it's intentional. I can add an alternative version that do that. but the problem is that if you only check 3 elements you can have elements that will never get into the function.

Reply:
What about if you have 2 elements and run it against 3 elements seq-map do you want to test the length of the list ot check if there are more then 3 elements? Or just call function on that two items?

Reply:
I think that best would be to create two versions, the user will decide which one he want.